service/src/test/java/bio/terra/pipelines/dependencies/leonardo/AppUtilsTest.java

```diff
@@ -0,0 +1,611 @@
+import org.broadinstitute.dsde.workbench.client.leonardo.model.ListAppResponse;
+import org.junit.jupiter.api.Test;
+
+class AppUtilsTest {
+  private final String workspaceId = UUID.randomUUID().toString();
+
+  private final LeonardoServerConfiguration leonardoServerConfiguration =
+      new LeonardoServerConfiguration(
+          "baseuri",
+          List.of("WDS", "CROMWELL"),
+          List.of("CROMWELL_RUNNER_APP", "CROMWELL"),
+          0,
+          false);
+
+  private final ListAppResponse separatedWdsAppWdsAppName;
+  // does not have "wds-" app name and does not have anticipated wds url
+  private final ListAppResponse separatedWdsAppNotWdsAppName;
+  private final ListAppResponse cromwellListAppResponse;
+  private final ListAppResponse cromwellListAppResponseProvisioning;
+  private final ListAppResponse cromwellListAppResponseNoProxyUrlDeleted;
+  private final ListAppResponse combinedWdsInCromwellApp;
+  private final ListAppResponse otherNamedCromwellAppOlder;
+  private final ListAppResponse galaxyApp;
+  private final ListAppResponse otherNamedCromwellApp;
+  private final ListAppResponse otherNamedCromwellAppProvisioning;
+  private final ListAppResponse separatedWorkflowsApp;
+
+  private final AppUtils au = new AppUtils(leonardoServerConfiguration);
+
+  @Test
+  void findWdsUrlInCombinedApp() {
+    List<ListAppResponse> apps = List.of(combinedWdsInCromwellApp);
+
+    AppUtils au = new AppUtils(leonardoServerConfiguration);
+    assertEquals(anticipatedWdsUrl("wds"), au.findUrlForWds(apps, workspaceId));
+  }
+
+  @Test
+  void findCromwellUrlInTerraApp() {
+    List<ListAppResponse> apps = List.of(cromwellListAppResponse);
+
+    AppUtils au = new AppUtils(leonardoServerConfiguration);
+    assertEquals(anticipatedCbasUrl("terra-app"), au.findUrlForCbas(apps, workspaceId));
+  }
+
+  @Test
+  void findWdsUrlInWdsApp() {
+    List<ListAppResponse> apps = List.of(separatedWdsAppWdsAppName);
+
+    AppUtils au = new AppUtils(leonardoServerConfiguration);
+    assertEquals(anticipatedWdsUrl("wds"), au.findUrlForWds(apps, workspaceId));
+  }
+
+  @Test
+  void findWdsAndCromwellInCombinedAppResponse() {
+    List<ListAppResponse> apps = List.of(separatedWdsAppWdsAppName, cromwellListAppResponse);
+
+    AppUtils au = new AppUtils(leonardoServerConfiguration);
+    assertEquals(anticipatedWdsUrl("wds"), au.findUrlForWds(apps, workspaceId));
+    assertEquals(anticipatedCbasUrl("terra-app"), au.findUrlForCbas(apps, workspaceId));
+  }
+
+  @Test
+  void preferSpecificallyNamedApp() {
+    List<ListAppResponse> apps =
+        new java.util.ArrayList<>(List.of(separatedWdsAppWdsAppName, separatedWdsAppNotWdsAppName));
+    permuteAndTest(apps, anticipatedWdsUrl("wds"));
+  }
+
+  @Test
+  void notChooseGalaxyApp() {
+    List<ListAppResponse> apps =
+        new java.util.ArrayList<>(List.of(otherNamedCromwellApp, galaxyApp));
+    // Shuffle to make sure the initial ordering isn't relevant:
+    Collections.shuffle(apps);
```

Comment:
is `shuffle()` deterministic? will it always change the order? (my cursory googling suggests it's random - couldn't that lead to inconsistent test results?)

Reply:
also why do you do this in some tests with multiple app responses but not others?

Reply:
huh no idea.  I thought all of these tests were using the `permuteAndTest` function which does a much nicer job than this shuffle thing.  will update

the tests using the shuffle (to be changed to permuteAndTest) are testing conditions where the order of the inputs matter like testing the `appComparisonFunction()`.  The tests that dont use shuffle are testing conditions where the order of the inputs dont matter

Reply:
thanks!